Droidcon-Boston/app/src/main/java/com/mentalmachines/droidcon_boston/views/agenda/AgendaDayViewModel.kt

```diff
@@ -0,0 +1,82 @@
+package com.mentalmachines.droidcon_boston.views.agenda
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.Transformations
+import androidx.lifecycle.ViewModel
+import com.google.firebase.database.DataSnapshot
+import com.google.firebase.database.DatabaseError
+import com.google.firebase.database.ValueEventListener
+import com.mentalmachines.droidcon_boston.data.FirebaseDatabase
+import com.mentalmachines.droidcon_boston.data.Schedule
+import com.mentalmachines.droidcon_boston.data.UserAgendaRepo
+import com.mentalmachines.droidcon_boston.firebase.FirebaseHelper
+import timber.log.Timber
+
+class AgendaDayViewModel(
+    private val dayFilter: String,
+    val onlyMyAgenda: Boolean,
+    private val userAgendaRepo: UserAgendaRepo
+) : ViewModel() {
+    private val firebaseHelper = FirebaseHelper.instance
+
+    private val _scheduleRows = MutableLiveData<List<Schedule.ScheduleRow>>()
+    private val _activeFilter = MutableLiveData<String>()
+
+    /**
+     * Whenever an active filter is set, we filter out the [_scheduleRows] for any that contain the
+     * filter within the title, description, or speaker names.
+     */
+    val scheduleRows: LiveData<List<Schedule.ScheduleRow>> = Transformations.map(_activeFilter) {
+            constraint ->
+        _scheduleRows.value?.filter { itemData ->
+            (itemData.talkTitle.contains(constraint, ignoreCase = true)
+                    || itemData.talkDescription.contains(constraint, ignoreCase = true)
+                    || itemData.speakerNames.any { it.contains(constraint, ignoreCase = true) })
+        }
+    }
+
+    private val dataListener: ValueEventListener = object : ValueEventListener {
+        override fun onDataChange(dataSnapshot: DataSnapshot) {
```

Comment:
to make this logic more testable, extract the `rows` variable management into its own method that returns `rows` when the relevant data from `dataSnapshot` gets passed in. 

I'm not sure how easy it is to test DataSnapshot objects which is why I'm recommending this. 

```
override fun onDataChange(dataSnapshot: DataSnapshot) {
    for (roomSnapshot in dataSnapshot.children) {
        val key = roomSnapshot.key ?: ""
        val data = roomSnapshot.getValue(FirebaseDatabase.ScheduleEvent::class.java)
        Timber.d("Event: $data")
        data.let {
            _scheduleRows.value = getRowsForData(key, data)
        }
    }
}
```
and in a separate method...

```
@VisibleForTesting fun getRowsForData(key: String, data: ScheduleEvent) {
    //.... etc
}
```

Reply:
I like this. I considered making the filtering method separate too, so I can write a unit test making sure that the filtering conditions are right. I can switch this up and add some tests.

Reply:
I refactored some stuff for testing, and the snapshots are testable I've pulled it off, but it's a little tricky and this PR already feels a little complicated. So I'd rather revisit it. 

Just pushed those changes but assuming CI passes let me know if you still think this is good.